### ISSUE TITLE
Haraka 2.8.28 installs incompatible haraka-plugin-redis version

### ISSUE THREAD

**Issue by DoobleD** [haraka/Haraka#3061]
I recently reinstalled Haraka's latest release ([2.8.28)](https://github.com/haraka/Haraka/releases/tag/2.8.28), along with a few plugins using the haraka-plugin-redis plugin. I was surprised to see some CRIT errors popping up on the karma plugin (and a few others when I tried removing the karma plugin):

```
[INFO] [-] [karma] connected to redis://undefined:undefined
[CRIT] [-] [server] Plugin karma timed out on hook init_master - make sure it calls the callback
```

After searching for a while, I realized the latest version of haraka-plugin-redis ([2.0.0](https://github.com/haraka/haraka-plugin-redis/releases/tag/2.0.0)) has recently [bumped](https://github.com/haraka/haraka-plugin-redis/pull/31) its node-redis dependency from 3.1.2 to 4.0.4. This is a breaking change. Reverting haraka-plugin-redis to 1.0.13 resolves the issue.

**Shouldn't the plugins latest releases all be compatible with Haraka's latest release?** Or at the very least, an Haraka release should only install plugin versions compatible with it.

**Comment by DoobleD** [comment]
```
$ git clone --branch 2.8.27 --single-branch https://github.com/haraka/Haraka.git && cd Haraka && npm install
$ npm install haraka-plugin-known-senders
...
$ npm list redis
├─┬ haraka-plugin-limit@1.0.5
│ └── redis@3.1.2 
├─┬ haraka-plugin-redis@2.0.0
│ └── redis@4.1.0 
├─┬ haraka-results@2.1.0
│ └── redis@3.1.2 
└── redis@3.0.2 
$ npm list haraka-plugin-redis
├─┬ haraka-plugin-karma@1.0.14
│ └── haraka-plugin-redis@2.0.0  deduped
├─┬ haraka-plugin-known-senders@1.0.7
│ └── haraka-plugin-redis@2.0.0  deduped
├─┬ haraka-plugin-limit@1.0.5
│ └── haraka-plugin-redis@2.0.0  deduped
├─┬ haraka-plugin-recipient-routes@1.0.2
│ └── haraka-plugin-redis@2.0.0  deduped
├── haraka-plugin-redis@2.0.0 
└─┬ haraka-plugin-watch@1.0.15
  └── haraka-plugin-redis@2.0.0  deduped
```

At least the plugins listed above are installing haraka-plugin-redis@2.0.0 (which installs redis@4.1.0) instead of haraka-plugin-redis@1.0.13.

**Comment by msimerson** [comment]
> Shouldn't the plugins latest releases all be compatible with Haraka's latest release? Or at the very least, an Haraka release should only install plugin versions compatible with it.

Yes, that certainly is the idea. That's why the redis and karma plugins have pinned version numbers in Haraka's package.json. It appears that where things went off the rails for you is when you did this: `npm install haraka-plugin-known-senders`. That plugin does not pin the version of haraka-plugin-redis. By manually installing that plugin, it upgraded your installed version of haraka-plugin-redis.

**Comment by DoobleD** [comment]
Thank you for the prompt feedback @msimerson! It appears the issue also occurs without installing any additional plugin:

```
$ git clone --branch 2.8.27 --single-branch https://github.com/haraka/Haraka.git && cd Haraka && npm install
...
$ npm list redis
├─┬ haraka-plugin-limit@1.0.5
│ └── redis@3.1.2 
├─┬ haraka-plugin-redis@2.0.0
│ └── redis@4.1.0 
├─┬ haraka-results@2.1.0
│ └── redis@3.1.2 
└── redis@3.0.2 
$ npm list haraka-plugin-redis
├─┬ haraka-plugin-karma@1.0.14
│ └── haraka-plugin-redis@2.0.0  deduped
├─┬ haraka-plugin-known-senders@1.0.7
│ └── haraka-plugin-redis@2.0.0  deduped
├─┬ haraka-plugin-limit@1.0.5
│ └── haraka-plugin-redis@2.0.0  deduped
├─┬ haraka-plugin-recipient-routes@1.0.2
│ └── haraka-plugin-redis@2.0.0  deduped
├── haraka-plugin-redis@2.0.0 
└─┬ haraka-plugin-watch@1.0.15
  └── haraka-plugin-redis@2.0.0  deduped
```

**Comment by msimerson** [comment]
I found the mismatch. We started pinning the version number of the plugin dependencies in PR #2965, which landed in 2.8.28. Prior Haraka versions, including 2.8.27, just had `*` and would, exactly as you demonstrated, install the most recent versions of any plugin.
